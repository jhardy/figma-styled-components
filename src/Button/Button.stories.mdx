import { Button } from './Button'
import { Text } from '../Text'
import { OptionButton } from '../OptionButton'
import { IconButton } from '../IconButton'
import { Story, Meta } from '@storybook/addon-docs/blocks'
import { Icon } from '../Icon/Icon'

<Meta title='MDX|Buttons' component={Button} />

# Buttons

<Story name='Buttons'>
  <div>
    <div className='component-box'>
      <Text size='xlarge' weight='bold'>
        Primary
      </Text>
      <br />
      <Text size='large'>Normal</Text>
      <div>
        <Button variant='primary'>Label</Button>
        <Button variant='primary' disabled={true}>
          Label
        </Button>
      </div>
      <Text size='large'>Destructive</Text>
      <div>
        <Button variant='primary' destructive>
          Label
        </Button>
        <Button variant='primary' disabled={true} destructive>
          Label
        </Button>
      </div>
    </div>
    <div className='component-box'>
      <Text size='xlarge' weight='bold'>
        Secondary
      </Text>
      <br />
      <Text size='large'>Normal</Text>
      <div>
        <Button>Label</Button>
        <Button disabled={true}>Label</Button>
      </div>
      <Text size='large'>Destructive</Text>
      <div>
        <Button destructive>Label</Button>
        <Button disabled={true} destructive>
          Label
        </Button>
      </div>
    </div>
     <div className='component-box'>
        <Text size='large'>Full Width</Text>
        <Button fullWidth={true}>Full Width</Button>
        <Button variant="primary" fullWidth={true}>Full Width</Button>
    </div>
  </div>
</Story>

<div className="props-table">

## Props

| Prop Name   | Type                                      | Default                  | Required | Description                           |
| :---------- | :---------------------------------------- | :----------------------- | :------: | :------------------------------------ |
| variant     | <code>"primary" &#124; "secondary"</code> | <code>"secondary"</code> |    -     | Button's appearance                   |
| destructive | boolean                                   | false                    |    -     | If button's action is destructive     |
| disabled    | boolean                                   | false                    |    -     | If button is disabled                 |
| fullWidth | boolean                                |    -    |    -     | Button takes up 100% of its container |
| onClick     | function                                  | -                        |    -     | Callback for Button's `onClick` event |

</div>

# Icon Button

<Story name='IconButton'>
  <div className='component-box'>
    <IconButton icon={<Icon name='Styles' />} />
  </div>
</Story>

<div className="props-table">

## Props

| Prop Name | Type                                   | Default | Required | Description                           |
| :-------- | :------------------------------------- | :-----: | :------: | :------------------------------------ |
| icon      | <code>JSX.Element &#124; string</code> |    -    |   yes    | Icon for the button                   |
| on        | boolean                                |  false  |    -     | If the button is toggled on           |
| disabled  | boolean                                |  false  |    -     | If the button is disabled             |
| onClick   | function                               |    -    |    -     | Callback for button's `onClick` event |

</div>

# OptionButton

<Story name='OptionButton'>
  <div className='component-box'>
    <OptionButton icon={<Icon name='LinkConnected' size={32} />} />
  </div>
</Story>

<div className="props-table">

## Props

| Prop Name | Type                                   | Default | Required | Description                           |
| :-------- | :------------------------------------- | :-----: | :------: | :------------------------------------ |
| icon      | <code>JSX.Element &#124; string</code> |    -    |   yes    | Icon for the button                   |
| on        | boolean                                |  false  |    -     | If the button is toggled on           |
| disabled  | boolean                                |  false  |    -     | If the button is disabled             |
| onClick   | function                               |    -    |    -     | Callback for button's `onClick` event |

</div>
