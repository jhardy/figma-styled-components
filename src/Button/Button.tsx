
import styled, { css } from 'styled-components'

export type ButtonVariants = 'primary' | undefined

export interface ButtonProps {
  variant?: ButtonVariants
  destructive?: boolean
  disabled?: boolean
}

const getButtonStyles = (
  variant: ButtonVariants,
  disabled: boolean
) => {
  switch (variant) {
    case 'primary':
      return css`
        color: #ffffff;
        background-color: ${disabled ? 'rgba(0, 0, 0, 0.3)' : '#18a0fb'};
        font-family: "Inter", sans-serif;
        line-height: 16px;
        font-weight: 500;
        font-size: 11px;
        line-height: 16px;
        letter-spacing: 0.01em;

        &:active,
        &:focus {
          box-shadow: inset 0 0 0 2px rgba(0, 0, 0, 0.3);
        }
      `
    default:
      return css`
        color: ${disabled ? 'rgba(0, 0, 0, 0.3)' : 'rgba(0, 0, 0, 0.8)'};
        box-shadow: ${disabled
          ? 'inset 0 0 0 1px rgba(0, 0, 0, 0.3)'
          : 'inset 0 0 0 1px rgba(0, 0, 0, 0.8)'};
        background-color: #ffffff;
        font-family: "Inter", sans-serif;
        font-weight: 500;
        font-size: 11px;
        line-height: 16px;
        letter-spacing: 0.005em;

        &:active,
        &:focus {

          box-shadow: inset 0 0 0 2px #18a0fb;
        }
      `
  }
}



const getDestructiveStyles = (variant: ButtonVariants) => {
  switch (variant) {
    case 'primary':
      return css`
        color: #ffffff;
        background-color: #f24822;

        &:active,
        &:focus {
          box-shadow: inset 0 0 0 2px rgba(0, 0, 0, 0.3);
        }
      `
    default:
      return css`
        color: #f24822;
        box-shadow: inset 0 0 0 1px #f24822;
        background-color: #ffffff;
        font-family: "Inter", sans-serif;
        font-weight: 500;
        font-size: 11px;
        line-height: 16px;
        letter-spacing: 0.005em;

        &:active,
        &:focus {
          box-shadow: inset 0 0 0 2px #f24822;
        }
      `
  }
}


export const Button = styled.button<ButtonProps>`
  box-sizing: border-box;
  align-items: center;
  display: inline-flex;
  justify-content: center;
  height: 30px;
  flex-shrink: 0;
  margin: 1px 0 1px 0;
  padding: 6px 16px 5px 16px;
  border: none;
  border-radius: 6px;
  outline: none;
  ${(props) => getButtonStyles(props.variant, props.disabled || false)};
  ${(props) =>
    props.destructive ? getDestructiveStyles(props.variant) : undefined}
  opacity: ${(props) => (props.disabled && props.destructive ? 0.4 : 1)};
`
