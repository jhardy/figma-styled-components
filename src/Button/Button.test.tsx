import 'jest-styled-components'
import * as React from 'react'
import * as renderer from 'react-test-renderer'

import { Button } from './Button'


test('A default Button', () => {
  const tree = renderer.create(<Button />).toJSON()
  expect(tree).toMatchSnapshot()
})

test('A primary Button', () => {
  const tree = renderer.create(<Button variant='primary' />)
  expect(tree).toMatchSnapshot()
})

test('A destructive Button', () => {
  const tree = renderer.create(<Button destructive={true} />)
})


test('A disabped Button', () => {
  const tree = renderer.create(<Button disabled={true} />).toJSON()
  expect(tree).toMatchSnapshot()
})

test('A full width Button', () => {
  const tree = renderer.create(<Button fullWidth={true} />).toJSON()
  expect(tree).toMatchSnapshot()
})



