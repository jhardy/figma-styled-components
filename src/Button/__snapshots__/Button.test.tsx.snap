// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A default Button 1`] = `
.c0 {
  box-sizing: border-box;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  height: 30px;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  margin: 1px 0 1px 0;
  padding: 6px 16px 5px 16px;
  border: none;
  border-radius: 6px;
  outline: none;
  color: rgba(0,0,0,0.8);
  box-shadow: inset 0 0 0 1px rgba(0,0,0,0.8);
  background-color: #ffffff;
  font-family: "Inter",sans-serif;
  font-weight: 500;
  font-size: 11px;
  line-height: 16px;
  -webkit-letter-spacing: 0.005em;
  -moz-letter-spacing: 0.005em;
  -ms-letter-spacing: 0.005em;
  letter-spacing: 0.005em;
  opacity: 1;
}

.c0:active,
.c0:focus {
  box-shadow: inset 0 0 0 2px #18a0fb;
}

<button
  className="c0"
/>
`;

exports[`A disabped Button 1`] = `
.c0 {
  box-sizing: border-box;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  height: 30px;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  margin: 1px 0 1px 0;
  padding: 6px 16px 5px 16px;
  border: none;
  border-radius: 6px;
  outline: none;
  color: rgba(0,0,0,0.3);
  box-shadow: inset 0 0 0 1px rgba(0,0,0,0.3);
  background-color: #ffffff;
  font-family: "Inter",sans-serif;
  font-weight: 500;
  font-size: 11px;
  line-height: 16px;
  -webkit-letter-spacing: 0.005em;
  -moz-letter-spacing: 0.005em;
  -ms-letter-spacing: 0.005em;
  letter-spacing: 0.005em;
  opacity: 1;
}

.c0:active,
.c0:focus {
  box-shadow: inset 0 0 0 2px #18a0fb;
}

<button
  className="c0"
  disabled={true}
/>
`;

exports[`A full width Button 1`] = `
.c0 {
  box-sizing: border-box;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  height: 30px;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  margin: 1px 0 1px 0;
  padding: 6px 16px 5px 16px;
  border: none;
  border-radius: 6px;
  outline: none;
  color: rgba(0,0,0,0.8);
  box-shadow: inset 0 0 0 1px rgba(0,0,0,0.8);
  background-color: #ffffff;
  font-family: "Inter",sans-serif;
  font-weight: 500;
  font-size: 11px;
  line-height: 16px;
  -webkit-letter-spacing: 0.005em;
  -moz-letter-spacing: 0.005em;
  -ms-letter-spacing: 0.005em;
  letter-spacing: 0.005em;
  opacity: 1;
  width: 100%;
}

.c0:active,
.c0:focus {
  box-shadow: inset 0 0 0 2px #18a0fb;
}

<button
  className="c0"
/>
`;

exports[`A primary Button 1`] = `
.c0 {
  box-sizing: border-box;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  height: 30px;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  margin: 1px 0 1px 0;
  padding: 6px 16px 5px 16px;
  border: none;
  border-radius: 6px;
  outline: none;
  color: #ffffff;
  background-color: #18a0fb;
  font-family: "Inter",sans-serif;
  line-height: 16px;
  font-weight: 500;
  font-size: 11px;
  line-height: 16px;
  -webkit-letter-spacing: 0.01em;
  -moz-letter-spacing: 0.01em;
  -ms-letter-spacing: 0.01em;
  letter-spacing: 0.01em;
  opacity: 1;
}

.c0:active,
.c0:focus {
  box-shadow: inset 0 0 0 2px rgba(0,0,0,0.3);
}

<button
  className="c0"
/>
`;
